components/style/traversal.rs

```diff
@@ -350,41 +471,31 @@ fn preprocess_children<'a, E, C, D>(context: &'a C,
             None => continue,
         };
 
-        // Set up our lazy child restyle data.
-        let mut child_data = unsafe { LazyRestyleData::<E, D>::new(&child) };
+        let mut child_data = unsafe { D::ensure_element_data(&child).borrow_mut() };
+        if child_data.is_unstyled_initial() {
+            continue;
+        }
+
+        let mut restyle_data = match child_data.restyle() {
+            Some(d) => d,
+            None => continue,
+        };
 
         // Propagate the parent and sibling restyle hint.
         if !propagated_hint.is_empty() {
-            child_data.ensure().map(|d| d.hint.insert(&propagated_hint));
+            restyle_data.hint.insert(&propagated_hint);
         }
 
-        // Handle element snashots.
-        if child_data.has_snapshot() {
-            // Compute the restyle hint.
-            let mut restyle_data = child_data.ensure().unwrap();
-            let mut hint = context.shared_context().stylist
-                                  .compute_restyle_hint(&child,
-                                                        restyle_data.snapshot.as_ref().unwrap(),
-                                                        child.get_state());
-
-            // If the hint includes a directive for later siblings, strip
-            // it out and modify the base hint for future siblings.
-            if hint.contains(RESTYLE_LATER_SIBLINGS) {
-                hint.remove(RESTYLE_LATER_SIBLINGS);
-                propagated_hint.insert(&(RESTYLE_SELF | RESTYLE_DESCENDANTS).into());
-            }
-
-            // Insert the hint.
-            if !hint.is_empty() {
-                restyle_data.hint.insert(&hint.into());
-            }
+        // Handle element snapshots.
+        let stylist = &context.shared_context().stylist;
+        let later_siblings = restyle_data.expand_snapshot(child, stylist);
+        if later_siblings {
+            propagated_hint.insert(&(RESTYLE_SELF | RESTYLE_DESCENDANTS).into());
         }
 
-        // If we restyled this node, conservatively mark all our children as
-        // needing a re-cascade. Once we have the rule tree, we will be able
-        // to distinguish between re-matching and re-cascading.
-        if restyled_parent {
-            child_data.ensure();
+        // If the parent's inherited properties changed, we need to recascade.
+        if parent_inherited_style_changed {
```

Comment:
Note that this check also needs to include whether we inherited any reset properties, so probably add a TODO for that, or similar.